Specmatic Config Revamp

sukesh2000 opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**: 
1. The configuration format is being updated to a newer version 

**Old version (Version 1):**
```yaml
version: 1
contract_repositories:
  - type: git
    repository: https://contracts
    provides:
      - com/petstore/1.yaml
    consumes:
      - com/petstore/payment.yaml

  - type: filesystem
    directory: contracts
    provides:
      - com/petstore/1.yaml
    consumes:
      - com/petstore/payment.yaml
      - com/petstore/order.yaml
```

**New version (Version 2):**
```yaml
version: 2
contracts:
  - git:
      url: https://contracts
      branch: 1.0.1
    provides:
      - product_search_bff_v4.yaml
    consumes:
      - api_order_v3.yaml
      - kafka.yaml

  - filesystem:
      directory: "specmatic-order-contracts"
    provides:
      - product_search_bff_v4.yaml
    consumes:
      - api_order_v3.yaml
      - kafka.yaml
```
2. A versioning system for the Specmatic configuration is introduced. The `contracts` key will be part of **version 2** of the configuration.  
3. A command-line tool is added to upgrade older versions of the configuration to the latest version (version 2).

<!-- Why are these changes necessary? -->

**Why**: 
1. Updating the `sources` key eliminates ambiguity and clarifies the purpose of its sub-keys.  
2. Introducing versioning enables a phased rollout of configuration updates without breaking existing applications that rely on older configuration versions.

<!-- How were these changes implemented? -->

**How**:
1. The **Adapter Pattern** was used to implement versioning.  
2. Two classes were introduced to represent different versions of the configuration:  
   - `SpecmaticConfigV1`: Represents the old configuration.  
   - `SpecmaticConfigV2`: Represents the new configuration.  
3. After reading the configuration file, the version value determines mapping to the appropriate version-specific class.  
   - These classes are then transformed into `SpecmaticConfig`, a unified class used across the application.  
   - The mapping is managed by the `VersionAwareConfigParser`.  
5. A command-line tool was created to handle configuration upgrades:  
   - **Command:** `specmatic config upgrade --input=<input-file> --output=<output-file>`  
   - If no input or output file is provided, it searches for the default configuration file (`specmatic.json`, `specmatic.yml`, or `specmatic.yaml`) and logs the output to the console.

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [x] Tests
- [ ] Sonar Quality Gate

<!-- Kindly link the issue that this PR will be addressing -->
<!-- **Issue ID**:
Closes: #123 -->

<!-- feel free to add additional comments -->

sukesh2000 committed:
Add version 2 for Specmatic Config

sukesh2000 committed:
Update version 2 of Specmatic Config, move config mapper to package and add version 1 for specmatic config

sukesh2000 committed:
Update error messages for Specmatic Config

sukesh2000 committed:
Add functionality to migrate versions

sukesh2000 committed:
Refactor SpecmaticConfig and update mapper name

sukesh2000 committed:
Add tests for SpecmaticConfig V1, V2 and factory

sukesh2000 committed:
Move SpecmaticConfig and other related data classes to previous state

sukesh2000 committed:
Move SpecmaticConfigTest to core package

sukesh2000 committed:
Rename SpecmaticConfigFactory to VersionAwareConfigParser and moved current implementation to extension functions

harikrishnan83 committed:
Setting default version 1 when version is not available, throw exception when unsupported version is provided, adding tests and constants for version numbers, added test specific config files

harikrishnan83 committed:
Fixing test by setting Specmatic config version to 1

sukesh2000 committed:
Move MigrateConfigCommand to a subcommand ConfigCommand

sukesh2000 committed:
Refactor SpecmaticConfigV2Test

yogeshnikam671 commented on application/src/main/kotlin/application/ConfigCommand.kt:
```diff
@@ -0,0 +1,92 @@
+@Command(
+    name = "config",
+    mixinStandardHelpOptions = true,
+    description = ["Manage and configure Specmatic Config."],
```
Let's call it Specmatic Configuration

sukesh2000 replied:
Done, renamed this at multiple places in the file

yogeshnikam671 commented on application/src/main/kotlin/application/ConfigCommand.kt:
```diff
@@ -0,0 +1,92 @@
+        @Option(names = ["--output"], description = ["Output file path for the migrated config."])
+        var outputFile: File? = null
+
+        private val latestVersion = SPECMATIC_CONFIG_VERSION_2
```
There is a coupling issue here. 
i.e. If you introduce a new config, you have to make sure to update this latestVersion here.

Let's decouple this by creating an enum class SpecmaticConfigVersion and access all the versions from this enum.
The latest version will always be - SpecmaticConfigVersion.values().maxByOrNull { it.version }

With this, this coupling will be removed and we will no longer need to explicitly update version over here.

sukesh2000 replied:
Done, created an enum class SpecmaticConfigVersion and now it's being used at all the places

yogeshnikam671 commented on application/src/main/kotlin/application/ConfigCommand.kt:
```diff
@@ -0,0 +1,92 @@
+
+        override fun call(): Int {
+            inputFile = inputFile ?: getConfigFile() ?: run {
+                logger.log("Missing input file. Please provide path to Specmatic Config.")
```
Unclear error message. It should be - "Default configuration file named specmatic.yaml/specmatic.yml/specmatic.json not found. Please provide the valid configuration file path using --input option."

sukesh2000 replied:
Updated the message as requested

yogeshnikam671 commented on application/src/main/kotlin/application/ConfigCommand.kt:
```diff
@@ -0,0 +1,92 @@
+            }
+
+            if (inputFile!!.readText().getVersion() == latestVersion) {
+                logger.log("Config is already up-to-date")
```
Message could be "The provided configuration file is already up-to-date."

sukesh2000 replied:
Updated the message as requested

yogeshnikam671 commented on application/src/main/kotlin/application/ConfigCommand.kt:
```diff
@@ -0,0 +1,92 @@
+                return FAILURE_EXIT_CODE
+            }
+
+            if (inputFile!!.readText().getVersion() == latestVersion) {
```
Let's avoid use !! operator.

sukesh2000 replied:
Removed it

yogeshnikam671 commented on application/src/main/kotlin/application/ConfigCommand.kt:
```diff
@@ -0,0 +1,92 @@
+        private val latestVersion = SPECMATIC_CONFIG_VERSION_2
+
+        override fun call(): Int {
+            inputFile = inputFile ?: getConfigFile() ?: run {
```
Bad idea to update the existing mutable variable `inputFile`. Create an immutable variable instead and store the result there.

sukesh2000 replied:
Done, I'm using a local variable for this purpose

yogeshnikam671 commented on application/src/main/kotlin/application/ConfigCommand.kt:
```diff
@@ -0,0 +1,92 @@
+            return SUCCESS_EXIT_CODE
+        }
+
+        private fun getConfigFile(): File? {
```
No need to create this function. Instead use the existing `getConfigFilePath` function which does the same thing and is already tested.

sukesh2000 replied:
Yes, using that now and removed `getConfigFile()` function

yogeshnikam671 commented on application/src/main/kotlin/application/ConfigCommand.kt:
```diff
@@ -0,0 +1,92 @@
+
+            if (outputFile != null) {
+                logger.log("Writing migrated config file to ${outputFile!!.path}")
+                outputFile!!.writeText(migratedConfigYaml)
```
Avoid using !!. You can use `let` operator instead. Also, wrap the writing logic in a try-catch block

sukesh2000 replied:
Now outputFile is a val so I've removed `!!` operator, and wrapped the whole block of the command's run method inside try catch so this is getting handled by that try catch block

yogeshnikam671 commented on application/src/main/kotlin/application/ConfigCommand.kt:
```diff
@@ -0,0 +1,92 @@
+        @Option(names = ["--input"], description = ["Path to config file that needs to migrated."])
+        var inputFile: File? = null
+
+        @Option(names = ["--output"], description = ["Output file path for the migrated config."])
```
Update the description - "File to write the updated configuration to. If not provided, the configuration will be logged in the console."

sukesh2000 replied:
Updated this

yogeshnikam671 commented on core/src/main/kotlin/io/specmatic/core/SpecmaticConfig.kt:
```diff
@@ -158,6 +155,77 @@ data class SpecmaticConfig(
+        val objectMapper = ObjectMapper(YAMLFactory()).registerKotlinModule()
+        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_DEFAULT)
+        return when (version) {
+            2 -> objectMapper.writeValueAsString(transformToV2())
```
Replace 2 with the version from the SpecmaticConfigVersion enum that needs to be created.

sukesh2000 replied:
Done

yogeshnikam671 commented on core/src/main/kotlin/io/specmatic/core/SpecmaticConfig.kt:
```diff
@@ -158,6 +155,77 @@ data class SpecmaticConfig(
+        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_DEFAULT)
+        return when (version) {
+            2 -> objectMapper.writeValueAsString(transformToV2())
+            else -> objectMapper.writeValueAsString(transformToV1())
```
Should not transform to V1 in the else branch. It should instead throw some exception

sukesh2000 replied:
Done, throwing `IllegalArgumentException` with a message

yogeshnikam671 commented on core/src/main/kotlin/io/specmatic/core/SpecmaticConfig.kt:
```diff
@@ -158,6 +155,77 @@ data class SpecmaticConfig(
+
+    private fun transformToV1(): SpecmaticConfigV1 {
+        return SpecmaticConfigV1(
+            version = 1,
```
Replace 1 with the enum value

sukesh2000 replied:
Done

yogeshnikam671 commented on core/src/main/kotlin/io/specmatic/core/SpecmaticConfig.kt:
```diff
@@ -158,6 +155,77 @@ data class SpecmaticConfig(
+
+    private fun transformToV2(): SpecmaticConfigV2 {
+        return SpecmaticConfigV2(
+            version = 2,
```
Replace 2 with the enum value

sukesh2000 replied:
Done

yogeshnikam671 commented on core/src/main/kotlin/io/specmatic/core/SpecmaticConfig.kt:
```diff
@@ -158,6 +155,77 @@ data class SpecmaticConfig(
+        return SpecmaticConfigV2(
+            version = 2,
+            contracts = this.sources.map { source ->
+                ContractConfig(
```
Instead of writing all this logic here. Write a static method in ContractConfig (e.g. contractConfigFrom(source)) and move this logic over there.

sukesh2000 committed:
Rename Migrate to Upgrade

yogeshnikam671 commented on core/src/main/kotlin/io/specmatic/core/config/VersionAwareConfigParser.kt:
```diff
@@ -0,0 +1,29 @@
+}
+
+fun String.getVersion(): Int {
+    return objectMapper.readTree(this).get(SPECMATIC_CONFIG_VERSION)?.asInt() ?: SPECMATIC_CONFIG_VERSION_1
```
Needs fixing. Behaves differently with JSON and YAML and if we explicitly add version as null

sukesh2000 replied:
I tested this and it's working fine as of now, please verify once and let me know

yogeshnikam671 commented on core/src/test/kotlin/io/specmatic/core/config/VersionAwareConfigParserTest.kt:
```diff
@@ -0,0 +1,60 @@
+    @Test
+    fun `should throw error when unsupported version is provided`() {
+        val specmaticConfigWithInvalidVersion =
+            File("./src/test/resources/specmaticConfigFiles/specmatic_without_unsupported_version.yaml")
```
spec file being used and the test name contradict with each other

sukesh2000 replied:
Changed the name of the test file

yogeshnikam671 commented on core/src/test/kotlin/io/specmatic/core/config/VersionAwareConfigParserTest.kt:
```diff
@@ -0,0 +1,60 @@
+import org.junit.jupiter.params.provider.CsvSource
+import java.io.File
+
+class VersionAwareConfigParserTest {
```
Parameterise all the tests to use both json and yaml formatted configs

sukesh2000 replied:
Done, created new JSON files for all the scenarios and I'm now testing on them as well

yogeshnikam671 commented on core/src/test/kotlin/io/specmatic/core/config/v1/SpecmaticConfigV1Test.kt:
```diff
@@ -0,0 +1,34 @@
+
+        assertThat(specmaticConfig.version).isEqualTo(specmaticConfigV1.version)
+        assertThat(specmaticConfig.sources.size).isEqualTo(specmaticConfigV1.sources.size)
+        assertThat(specmaticConfig.sources[0].test).containsAll(specmaticConfigV1.sources[0].test)
```
sources being a data class you should be able to directly do the equality check among the two.

sukesh2000 replied:
Yes, updated the tests to do equality check on sources directly

sukesh2000 committed:
Refactor ConfigCommand as per review comments and add SpecmaticConfigVersion enum

sukesh2000 committed:
Refactor classes to use SpecmaticConfigVersion enum

sukesh2000 committed:
Refactor SpecmaticConfig to do the conversion from Source to ContractConfig in ContractConfig class

sukesh2000 committed:
Refactor tests to test on JSON and YAML files of the config

sukesh2000 committed:
Refactor SpecmaticConfigV1Test

yogeshnikam671 commented on core/src/main/kotlin/io/specmatic/core/SpecmaticConfig.kt:
```diff
@@ -158,6 +153,43 @@ data class SpecmaticConfig(
+        objectMapper.setSerializationInclusion(JsonInclude.Include.NON_DEFAULT)
+        return when (version) {
+            SpecmaticConfigVersion.VERSION_2 -> objectMapper.writeValueAsString(upgradeToV2())
+            else -> throw IllegalArgumentException(
```
Where is the SpecmaticConfigVersion.VERSION_1 branch? Let's keep it for consistency even though it won't be used currently.

sukesh2000 replied:
As Hari said we don't need to any upgrade function for version 1, so there won't be any need for version 1 branch

yogeshnikam671 commented on core/src/main/kotlin/io/specmatic/core/config/v2/SpecmaticConfigV2.kt:
```diff
@@ -0,0 +1,55 @@
+) {
+    fun transform(): SpecmaticConfig {
+        return SpecmaticConfig(
+            version = this.version,
```
Inconsistent. We should have similar assignment of version using enum instead of this.version.

sukesh2000 replied:
Done

yogeshnikam671 commented on core/src/test/kotlin/io/specmatic/core/config/v2/SpecmaticConfigV2Test.kt:
```diff
@@ -0,0 +1,74 @@
+        val specmaticConfig = specmaticConfigV2.transform()
+
+        assertThat(specmaticConfig.version).isEqualTo(specmaticConfigV2.version)
+        assertThat(specmaticConfig.sources.size).isEqualTo(specmaticConfigV2.contracts.size)
```
Do similar refactoring as what's been done in VersionAwareConfigParserTests

sukesh2000 replied:
This won't be possible as Sources and ContractConfig are different classes

yogeshnikam671 commented on core/src/test/kotlin/io/specmatic/core/config/v2/SpecmaticConfigV2Test.kt:
```diff
@@ -0,0 +1,74 @@
+        val specmaticConfig = specmaticConfigV2.transform()
+
+        assertThat(specmaticConfig.version).isEqualTo(specmaticConfigV2.version)
+        assertThat(specmaticConfig.sources.size).isEqualTo(specmaticConfigV2.contracts.size)
```
Do similar refactoring as to what's been done in VersionAwareConfigParserTests

sukesh2000 replied:
This won't be possible as Sources and ContractConfig are different classes

yogeshnikam671 committed:
Refactor ConfigCommand and some tests

sukesh2000 committed:
Refactor SpecmaticConfig to use SpecmaticConfigVersion data type instead of Int and address review comments

sukesh2000 committed:
Refactor config command to upgrade only to the latest version of the config

sukesh2000 committed:
Update Push & Subscribe command to use SpecmaticConfig instead of JSON Object

nashjain committed:
cleaned up the tests

sukesh2000 committed:
Move upgrade functionality from SpecmaticConfig to ConfigCommand

sukesh2000 committed:
Remove changes related to Push & Subscribe command

sukesh2000 committed:
Remove unecessary changes

sukesh2000 committed:
Refactor Contract Config

sukesh2000 committed:
Add tests for Contract Config serializer & deserializer

sukesh2000 committed:
Merge branch 'main' of https://github.com/znsio/specmatic into specmatic-config-revamp

nashjain committed:
moved the getLatestConfig() from ConfigCommand to the respective versioned config class, so that we don't break SRP
Created 2 interfaces: SpecmaticVersionedConfig and SpecmaticVersionedConfigLoader to avoid depending on specific version details, instead talk to the interfaces (avoid breaking DIP)
Added a method convertToLatestVersionedConfig() to SpecmaticConfigVersion, to ensure that in future when we've new versions, we don't need to go and change code in multiple places (avoid breaking OCP)
Removed the dependency on org.skyscreamer:jsonassert

nashjain committed:
Replaced conditional logic with polymorphism
ContractSource does not need to be a sealed class, we can replace it with interface
